### ISSUE TITLE
Fix NPN guard for OpenSSL 1.1

### ISSUE THREAD

**Issue by d101d8a3-23b1-451b-ad28-9b9c30955bad** [python/cpython#74807]
BPO | [30622](https://bugs.python.org/issue30622)
--- | :---
Nosy | @tiran, @ned-deily, @melvyn-sopacua
PRs | <li>python/cpython#2079</li><li>python/cpython#3314</li><li>python/cpython#3316</li><li>python/cpython#5341</li><li>python/cpython#5343</li><li>python/cpython#5859</li><li>python/cpython#5860</li><li>python/cpython#5861</li><li>python/cpython#5863</li><li>python/cpython#5876</li><li>python/cpython#5880</li><li>python/cpython#5881</li><li>python/cpython#5882</li><li>python/cpython#6102</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = 'https://github.com/tiran'
closed_at = <Date 2017-09-05.14:06:39.186>
created_at = <Date 2017-06-10.15:08:14.661>
labels = ['expert-SSL', 'build', '3.7']
title = 'Fix NPN guard for OpenSSL 1.1'
updated_at = <Date 2018-03-28.18:27:17.332>
user = 'https://github.com/melvyn-sopacua'
```

bugs.python.org fields:
```python
activity = <Date 2018-03-28.18:27:17.332>
actor = 'ned.deily'
assignee = 'christian.heimes'
closed = True
closed_date = <Date 2017-09-05.14:06:39.186>
closer = 'christian.heimes'
components = ['SSL']
creation = <Date 2017-06-10.15:08:14.661>
creator = 'msopacua'
dependencies = []
files = []
hgrepos = []
issue_num = 30622
keywords = []
message_count = 15.0
messages = ['295643', '301268', '301281', '301312', '301313', '312757', '312762', '312788', '312789', '312803', '312807', '312811', '312812', '313811', '314616']
nosy_count = 3.0
nosy_names = ['christian.heimes', 'ned.deily', 'msopacua']
pr_nums = ['2079', '3314', '3316', '5341', '5343', '5859', '5860', '5861', '5863', '5876', '5880', '5881', '5882', '6102']
priority = 'normal'
resolution = 'fixed'
stage = 'resolved'
status = 'closed'
superseder = None
type = 'compile error'
url = 'https://bugs.python.org/issue30622'
versions = ['Python 2.7', 'Python 3.6', 'Python 3.7']
```

</p></details>

**Comment by d101d8a3-23b1-451b-ad28-9b9c30955bad** [comment]
Starting OpenSSL 1.1, the variable OPENSSL_NPN_NEGOTIATED is always defined. This makes the NPN detection fail.

Related info: https://bugs.freebsd.org/bugzilla/show_bug.cgi?id=217703

**Comment by tiran** [comment]
New changeset b2d096bd2a5ff86e53c25d00ee5fa097b36bf1d8 by Christian Heimes (Melvyn Sopacua) in branch 'master':
bpo-30622: Change NPN detection: (bpo-2079)
https://github.com/python/cpython/commit/b2d096bd2a5ff86e53c25d00ee5fa097b36bf1d8

**Comment by tiran** [comment]
New changeset 72ed233167b10d3a488d30a8ec3a17e412a7dd69 by Christian Heimes in branch '2.7':
[2.7] bpo-30622: Change NPN detection: (GH-2079) (bpo-3316)
https://github.com/python/cpython/commit/72ed233167b10d3a488d30a8ec3a17e412a7dd69

**Comment by tiran** [comment]
New changeset 7316c6d4a57931e9786c06eae168b227d7463317 by Christian Heimes in branch '3.6':
[3.6] bpo-30622: Change NPN detection: (GH-2079) (bpo-3314)
https://github.com/python/cpython/commit/7316c6d4a57931e9786c06eae168b227d7463317

**Comment by tiran** [comment]
Thanks for your patch!

I have merged it into master, 3.6, and 2.7. The other branches are in security fix-only mode.
